src/storage/store/rel_table_data.cpp

```diff
@@ -291,6 +308,160 @@ void RelTableData::prepareCommitCSRNGWithoutSliding(Transaction* transaction,
+                                    0;
+        int64_t numRowsAfterSlide = numRowsInCSR + numInsertions - numDeletions;
+        KU_ASSERT(numRowsAfterSlide >= 0);
+        currentCSROffset += numRowsAfterSlide;
+        slidedCSRChunk->setValue<offset_t>(currentCSROffset, i);
+    }
+    for (auto i = currentNumSrcNodesInNG; i < StorageConstants::NODE_GROUP_SIZE; i++) {
+        auto nodeOffset = nodeGroupStartOffset + i;
+        // Assert that should only have insertions for these nodes.
+        KU_ASSERT(!relNodeGroupInfo->deleteInfo.contains(nodeOffset));
+        for (auto columnID = 0u; columnID < columns.size(); columnID++) {
+            KU_ASSERT(!relNodeGroupInfo->updateInfoPerChunk[columnID].contains(nodeOffset));
+        }
+        auto numRowsInCSR = 0;
+        if (relNodeGroupInfo->adjInsertInfo.contains(nodeOffset)) {
+            // This node is inserted now. Update csr offset accordingly.
+            numRowsInCSR += relNodeGroupInfo->adjInsertInfo.at(nodeOffset).size();
+            newNumSrcNodesInNG = i + 1;
+        }
+        currentCSROffset += numRowsInCSR;
+        slidedCSRChunk->setValue<offset_t>(currentCSROffset, i);
+    }
+    slidedCSRChunk->setNumValues(newNumSrcNodesInNG);
+    return slidedCSRChunk;
+}
+
+std::unique_ptr<ColumnChunk> RelTableData::slideCSRColumnChunk(Transaction* transaction,
+    ColumnChunk* csrOffsetChunk, ColumnChunk* slidedCSROffsetChunkForCheck, ColumnChunk* relIDChunk,
+    const offset_to_offset_to_row_idx_t& insertInfo,
+    const offset_to_offset_to_row_idx_t& updateInfo, const offset_to_offset_set_t& deleteInfo,
+    node_group_idx_t nodeGroupIdx, Column* column, LocalVectorCollection* localChunk) {
+    auto oldCapacity = csrOffsetChunk->getNumValues() == 0 ?
+                           0 :
+                           csrOffsetChunk->getValue<offset_t>(csrOffsetChunk->getNumValues() - 1);
+    auto newCapacity = slidedCSROffsetChunkForCheck->getNumValues() == 0 ?
+                           0 :
+                           slidedCSROffsetChunkForCheck->getValue<offset_t>(
+                               slidedCSROffsetChunkForCheck->getNumValues() - 1);
+    // TODO: No need to allocate the new column chunk if this is relID.
+    auto columnChunk = ColumnChunkFactory::createColumnChunk(
+        column->getDataType()->copy(), enableCompression, oldCapacity);
+    column->scan(transaction, nodeGroupIdx, columnChunk.get());
+    auto newColumnChunk = ColumnChunkFactory::createColumnChunk(
+        column->getDataType()->copy(), enableCompression, newCapacity);
+    auto currentNumSrcNodesInNG = csrOffsetChunk->getNumValues();
+    auto csrOffsets = (offset_t*)csrOffsetChunk->getData();
+    auto relIDs = (offset_t*)relIDChunk->getData();
+    auto nodeGroupStartOffset = StorageUtils::getStartOffsetOfNodeGroup(nodeGroupIdx);
+    for (auto i = 0; i < currentNumSrcNodesInNG; i++) {
+        auto nodeOffset = nodeGroupStartOffset + i;
```

Comment:
You don't have to slide node one by one if that node doesn't have any updates.
You can accumulate nodes, until you encounter a node that has updates

Reply:
Separate this into another PR.